### PR TITLE
kernel: introduce KERNEL_WERROR config option

### PR DESCRIPTION
**Pull request by ynezz**
In commit b2d1eb717b65 ("generic: 5.15: enable Werror by default for kernel compile") CONFIG_WERROR=y was enabled and all warnings/errors reported with GCC 12 were fixed.

Keeping this in sync with past/future GCC versions is going to be uphill battle, so lets introduce new KERNEL_WERROR config option, enable it by default only for tested/known working combinations and on buildbots.

References: #12687 #12622 #12734

### CODE REVIEW

**Comment by neheb** [comment]
OTOH it being painful was the reason for upstream's introduction of Werror.

**Comment by robimarko** [comment]
Enabling it on GCC12 now only delays the same warnings appearing once GCC13 is the default, and as @neheb pointed out its supposed to be painful.

**Comment by ynezz** [comment]
> Enabling it on GCC12 now only delays the same warnings appearing once GCC13 is the default.

It would be spotted on buildbots as I've tied it to CONFIG_BUILDBOT option as well. I've suggested to enable `CONFIG_BUILDBOT=y` in the CI  as well, but it was rejected. I'm not sure if we want to add `CONFIG_CI=y` to enable checks for the CI only.

**Comment by robimarko** [comment]
My comment was more in a sense that option of disabling `WERROR` should not even exist.
Also, to me it makes more sense to have CI catch the warning first before buildbots even get a chance to build it.

**Comment by neheb** [comment]
@Ansuel has the most say as he introduced Werror and fixed most of the warnings. I guess I shouldn't mention GCC11 and 13 :)

**Comment by ynezz** [comment]
That bugreport #12687 triggered this change, I'm anticipating much more of such bug reports as this is enabled in upcoming stable 23.05 release.

> I still see making things fail and not compile the correct way to push people fixing their stuff

#### Expected behaviour
* Kernel should successfully build

IMO it says it all. Folks expect, that its going to work, they're not going to fix our stuff :)

> buildbot doesn't use it so I don't see a problem in having it currently broken

I understand, that its a huge amount of work to keep `gcc-11`, `gcc-12` and `gcc-13` in working state with such pedantic configuration, so that's why I've introduced that option to relax it a bit. IMO it doesn't need to be broken for end users, its bad user experience.

> Do we have some link on why it was rejected?

[Rejected is probably not a proper word](https://github.com/openwrt/openwrt/pull/10407#discussion_r940032758), lets rephrase it to _not considered_

**Comment by timocapa** [comment]
> even if you point them at the fix

ive not had time yet to properly look into how to make patches that could/would be accepted in openwrt, sorry.

i'll look at https://openwrt.org/docs/guide-developer/toolchain/use-patches-with-buildsystem in some time.

**Comment by ynezz** [comment]
> I'm anticipating much more of such bug reports as this is enabled in upcoming stable 23.05 release.

Slowly incoming https://github.com/openwrt/openwrt/issues/12734

**Comment by mpratt14** [comment]
I know that the intention is to make user experience better for a stable branch, but that in turn would make the stable branch "less stable" (provided that the specific warning that WERROR turns into an error is not useless).

Just throwing an idea out there, what if stable branches had it enabled by default, and master disabled by default? or maybe the reverse if we really think user experience is more important, master enabled by default, and stable branch disabled by default?

**Comment by robimarko** [comment]
Considering that master will eventually become next stable that just kicks the can down the road.

**Comment by mpratt14** [comment]
I'm more or less imagining a way for Make to know whether a stable branch is checked out or not and use an if statement. I don't know if it's possible.

**Comment by ynezz** [comment]
> I know that the intention is to make user experience better for a stable branch

Its bad for use experience, regardless of the branch or the environment. IIUC then original intention is QA purpose, which should be part of internal workflows, we shouldn't impose QA tasks on every single user out there.

The rule of thumb for this project is dead simple, if it breaks, fix or revert it. From the linked bug reports you can clearly see, that its broken for some users, so we should fix it.

Those bug reports always represent just a fraction of the ongoing issues, interpolating it reveals a lot more of such WTF moments.

Fixing the linked issues is not going to cut it, extending the CI matrix with another GCC version is not going to help either, because folks are simply allowed to use different combinations of kernel config options, use out of tree targets, kernel modules etc. so we can't simply CI QA all of this combinations and we're always going to spot something.

>  so lets introduce new KERNEL_WERROR config option, enable it by default only for tested/known working combinations and on buildbots.

In retrospect this was wrong as well, there is no such tested/known working combination, so `KERNEL_WERROR` option should be only enabled in our internal QA workflows, such as CI and buildbots.

**Comment by Ansuel** [comment]
@ynezz I totally understand the reason and you are absolutely right! What I see as a problem is the gcc thing (I would drop it, I get the idea that gcc12 works and every warning is fixed but I feel like that is a hint that we will have to had gcc13 when we will have things fixed, some kind of validation config? I don't like that part since we will have to remember adding stuff to it)

I like the idea of enabling it only in CI and buildbots. For CI we would just enable this new config option but I wonder if a good idea might be introduce a CI config that would enable all of this options? (another config that would be added is the AUTOREMOVE and also CCACHE?) I remember there was an idea so this might be a good time to implement it?

**Comment by Ansuel** [comment]
I just read the problem with using buildbot config... maybe we can find some common config to improve reproducibility between CI and buildbot?
